engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobBatchCollector.java

```diff
@@ -0,0 +1,169 @@
+          // as well as one byte required per job key for its type header. if we ever add more, this
+          // should be updated accordingly.
+          final var jobRecordLength = jobRecord.getLength();
+          final var expectedEventLength = record.getLength() + jobRecordLength + Long.BYTES + 1;
+          if (activatedCount.value <= maxActivatedCount
+              && canWriteEventOfLength.test(expectedEventLength)) {
+            appendJobToBatch(jobIterator, jobKeyIterator, jobCopyBuffer, key, jobRecord);
+            activatedCount.increment();
+          } else {
+            // if no jobs were activated, then the current job is simply too large, and we cannot
+            // activate it
+            if (activatedCount.value == 0) {
+              unwritableJob.set(new LargeJob(key, jobRecord));
+            }
+
+            value.setTruncated(true);
+            return false;
+          }
+
+          return activatedCount.value < maxActivatedCount;
+        });
+
+    if (unwritableJob.ref != null) {
+      return Either.left(unwritableJob.ref);
+    }
+
+    return Either.right(activatedCount.value);
+  }
+
+  private void setJobVariables(
+      final Collection<DirectBuffer> requestedVariables,
+      final JobRecord jobRecord,
+      final long elementInstanceKey) {
+    if (elementInstanceKey >= 0) {
+      final DirectBuffer variables = collectVariables(requestedVariables, elementInstanceKey);
+      jobRecord.setVariables(variables);
+    } else {
+      jobRecord.setVariables(DocumentValue.EMPTY_DOCUMENT);
+    }
+  }
+
+  private void appendJobToBatch(
+      final ValueArray<JobRecord> jobIterator,
+      final ValueArray<LongValue> jobKeyIterator,
+      final ExpandableArrayBuffer jobCopyBuffer,
+      final Long key,
+      final JobRecord jobRecord) {
+    jobKeyIterator.add().setValue(key);
+    final JobRecord arrayValueJob = jobIterator.add();
+
+    // clone job record since buffer is reused during iteration
+    jobRecord.write(jobCopyBuffer, 0);
+    arrayValueJob.wrap(jobCopyBuffer, 0, jobRecord.getLength());
+  }
+
+  private Collection<DirectBuffer> collectVariableNames(final JobBatchRecord batchRecord) {
+    final ValueArray<StringValue> requestedVariables = batchRecord.variables();
+
+    variableNames.clear();
+    requestedVariables.forEach(
+        variable -> variableNames.add(BufferUtil.cloneBuffer(variable.getValue())));
+
+    return variableNames;
+  }
+
+  private DirectBuffer collectVariables(
+      final Collection<DirectBuffer> variableNames, final long elementInstanceKey) {
+    final DirectBuffer variables;
+
+    if (variableNames.isEmpty()) {
+      variables = variableState.getVariablesAsDocument(elementInstanceKey);
+    } else {
+      variables = variableState.getVariablesAsDocument(elementInstanceKey, variableNames);
+    }
+
+    return variables;
+  }
+
+  record LargeJob(long key, JobRecord record) {}
```

Comment:
:thought_balloon: Unsure about this, but it seems like we should already have a way to group a job and it's key